### ISSUE TITLE
Validation success but predict all black images

### ISSUE THREAD

**Issue by Crystal-Biru** [MIC-DKFZ/nnUNet#1825]
I trained my model successful, and the validation is correct. But when I use the following command to predict:
`nnUNetv2_predict -i ./nnUNet_raw/Dataset160_ODOCSegmentation/imagesTs/ -o ./nnUNet_raw/Dataset160_ODOCSegmentation/imagesTsPred/ -d 160 -c 2d -f 0`
I found the output images are all black. I use imagesTr to test the prediction too, but the output images are also totally black. I am really curious about this and wondering how to solve it.
I use the following commands:
`nnUNetv2_plan_and_preprocess -d 160 --verify_dataset_integrity

nnUNetv2_train 160 2d 0 --npz

nnUNetv2_predict -i ./nnUNet_raw/Dataset160_ODOCSegmentation/imagesTs/ -o ./nnUNet_raw/Dataset160_ODOCSegmentation/imagesTsPred/ -d 160 -c 2d -f 0`
Only prediction made error.

**Comment by Crystal-Biru** [comment]
Thank you for your reply! I noticed that problem and checked the labels through python (I write a program to show the label through pyplot and output the non-zero values of label). When check the output label obtained from validation, they are shown to be well segmented, and value 1, 2 are outputted. However, when I check the output label obtained from prediction, they are shown to be totally black and no value is outputted. 
**The labels got from prediction are exactly empty, all values are 0. But the labels got from validation have values 0, 1, 2, and the segmentation performs well.** This makes me really confused.

**Comment by ykirchhoff** [comment]
Hi,

that should really not happen. There are a few things that might in principle go wrong, but nothing which would explain to me that the predictions are completely blank. Can you maybe run the predictions again with --npz? This saves the softmax outputs to .npz files and then you could at least check that the model is creating some - more or less - sensible output.
Otherwise, if that is possible for you, you can send me the dataset and trained model folder, so I can try to debug it on my machine.

Best,
Yannick

**Comment by Crystal-Biru** [comment]
I can send you the data set, and I would really appreciate it if you could try to debug it for me.  Could you please tell me which email I can send the dataset to?

**Comment by ykirchhoff** [comment]
Hi,

you can send it to yannick.kirchhoff@dkfz.de. If it is too large to send via mail I can also send you a link to upload the data.
Please mention the issue in the mail so I can directly link it.

Best,
Yannick

**Comment by Crystal-Biru** [comment]
I have already sent the dataset, have you received it?

**Comment by ykirchhoff** [comment]
Hi, yes I got it. It will take some time because I am on a conference next week but I will look into it when I am back.

Best,
Yannick

**Comment by ykirchhoff** [comment]
Hi, the link you sent me unfortunately expired. I just sent you an email with a link where you can upload the data again.
Best,
Yannick

**Comment by ykirchhoff** [comment]
Hi @Crystal-Biru,

I will close this issue for now to clean up open issues a little. Feel free to reopen the issue and send me the data again if your problem still persists.

Best,
Yannick

**Comment by RoloSmajda** [comment]
Hi @ykirchhoff ,
I have exactly the same problem as was described by @Crystal-Biru. I trained my model on amos dataset (well part of it). Validation predictions that are done at the end of fold training are what you would expect (although not completly correct, due to just training for 50 epochs), but predicting on trained model outputs empty segmentations. Also, when i try to open that predicted segmentation in ITK-SNAP, i get this error message:
`Error: exception occured during image IO. Exception: C:\Users\pyush\tk\buildbot\Nightly\itk\v5.2.1\itk\Modules\IO\NIFTI\src\itkl 1980: ITK ERROR: ITK only supports orthonormal direction cosines. No or definition found!`
I tried to look online for solution but got nowhere so far. Would appriciate any help or suggestions.

**Comment by ykirchhoff** [comment]
Hi @RoloSmajda,

that should definitely not happen. What data do you use to predict on? And can you open the test images with ITK-SNAP? Maybe they are already corrupted.

Best,
Yannick

**Comment by huangcq18** [comment]
I came across the same problem. Line 521 of inference/predict_from_raw_data.py `prediction = self.network(x)` produced 'nan'. I also notice it only fails on gpu while works successfully on cpu. I am very comfused.

**Comment by RoloSmajda** [comment]
> Hi @RoloSmajda,
> 
> that should definitely not happen. What data do you use to predict on? And can you open the test images with ITK-SNAP? Maybe they are already corrupted.
> 
> Best, Yannick

Thank you for reply Yannick. Yes i can open predicted data, its just normal CT scan, same as the ones that were used for training. I trained many folds and i always have the same problem. Im also confused by this behaviour, same as @huangcq18.

**Comment by ykirchhoff** [comment]
It is really weird that it works on CPU but fails on GPU. Can one or both of you share the trained models and data with me, so I can debug it locally? It is quite difficult to figure out what's going wrong without seeing it. Just send me an email to yannick.kirchhoff@dkfz.de and mention this issue, then I will share a link with you where you can upload everything.

**Comment by ykirchhoff** [comment]
Hi @RoloSmajda,

thanks for the model and the test cases. Unfortunately I am not able to reproduce your issue on my machine so we will need to go back to remote debugging. I attached one example prediction of one of the test cases below.
![amos_0026](https://github.com/MIC-DKFZ/nnUNet/assets/104990536/eb05ec24-ba8a-4953-926b-f56e4d8d0fcb)
Here are some more general things you can try before we dive deeper into debugging:
- Do you have the most recent version of nnUNet installed? And are all your dependencies up-to-date? Maybe something is broken there.
- There might be issues with your hardware, especially with your GPU. If you have another machine with a similar setup you could try reproducing your issue there. However, if it is related to the GPU I would usually expect to see issues already during training.
- If you are confident with the code you can check where approximately the problem appears. Specifically, do you get nans like @huangcq18 in the forward pass during prediction, i.e. in the line `prediction = self.network(x)`, and if so is the input x already broken? That might narrow down the search space quite a bit.

Let me know if you need help with any of these things.

Best,
Yannick

**Comment by RoloSmajda** [comment]
Thank you for reply.

I do not have another hardware to test it. 

So this is the input x : 
[[[[-7.9972, -7.9972, -7.9972,  ..., -7.9972, -7.9972, -7.9972],
          [-7.9972, -7.9972, -7.9972,  ..., -7.9972, -7.9972, -7.9972],
          [-7.9972, -7.9972, -7.9972,  ..., -7.9972, -7.9972, -7.9972],
          ...,
          it is of shape [1, 1, 512, 448].
          
This is the prediction after `prediction = self.network(x)` :
[[[[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          this is of shape [1, 16, 512, 448].
          
So yes, it is full of nans.

However, i trierd to predict on cpu and prediction looks reasonable:
tensor([[[[ 1.3745e+01,  1.5053e+01,  1.5855e+01,  ...,  1.6014e+01,
            1.5603e+01,  1.3954e+01],
          [ 1.4559e+01,  1.7675e+01,  1.7689e+01,  ...,  1.9225e+01,
            1.7207e+01,  1.5181e+01],
          [ 1.5202e+01,  1.7072e+01,  1.9358e+01,  ...,  1.8488e+01,
            1.8178e+01,  1.5178e+01],
            
The prediction works on cpu, i get segmentation mark as expected, so literally same issue as @huangcq18 .             
Makes me wonder if my cuda installation could be wrong?

**Comment by ykirchhoff** [comment]
Hi @RoloSmajda,

the input and CPU output both look reasonable. So the issue is indeed in the forward pass on the GPU.
It is definitely worth a try setting up a completely new environment from scratch. The only thing that confuses me a bit is that training worked without any issues. But let's give that a try and hope for the best.

Best,
Yannick

**Comment by ykirchhoff** [comment]
Hi @RoloSmajda,

I just thought of something, if you are setting up a new environment also check that your NVIDIA driver is up-to-date. We recently had some issues with incompatibilities of driver versions, which lead to different problems but was quite annoying.

Best,
Yannick

**Comment by ykirchhoff** [comment]
Hi @RoloSmajda,

did you already find the time to setup a new environment and test it?

Best,
Yannick

**Comment by RoloSmajda** [comment]
Hi @ykirchhoff,

no, not yet. I've been abroad for past few days. When i'll have time, i'll test it and let you know.

**Comment by ykirchhoff** [comment]
No worries, just wanted to check :)

**Comment by RoloSmajda** [comment]
Hi @ykirchhoff.

So i updated my drivers, installed new version of cuda (12.3) and created new enviroment. The issue is still persists. Im really at lost to what could be causing it.

**Comment by FabianIsensee** [comment]
Hey, so there is one small thing you could try. Your problem description doesn't quite match this fix but while we are grasping at straws we might as well try that. Last year we had a bug where predictions could in part contain nans which would then be mapped to 0 in the argmax. This was related to a fp16 overflow. Can you try installing from the current master branch (not `pip install nnunetv2`)?

**Comment by FabianIsensee** [comment]
@ykirchhoff so far we have not been able to reproduce this problem on our end, correct?
@RoloSmajda @huangcq18 what GPU did you use in your experiments?

**Comment by FabianIsensee** [comment]
Please also verify that there are no hidden nan values in your input by checking the output of `torch.any(torch.isnan(input_image))`

**Comment by ykirchhoff** [comment]
@FabianIsensee yes I was not able to reproduce the issue so far. But I also only tested with the most recent version, so if the pypi version is not the most recent one, it might be related to that (in case nnUNet was installed this way)

**Comment by RoloSmajda** [comment]
Hi @FabianIsensee, thank you, for also trying to help.

I already had installation from master branch, not via pip. To be sure, i pulled newest changes, but it didn't fix anything. 

The GPU im using is NVIDIA GeForce GTX 1650 Ti. I know, not great, but it managed to train, and then validate, so i'd expect it also to predict. 

I'm bit confused about this: `torch.any(torch.isnan(input_image))`. If you meant, just to load the file i'm trying to predict on, and test if there are any NaNs, then i did that and it was false, so no NaNs.

**Comment by FabianIsensee** [comment]
What I mean is the 'x' from `prediction = self.network(x)`. Just wanted to make sure that there are no NaNs in the input which could propagate to the output.
What is really curious is that we cannot reproduce this problem on our end, even with the same image and checkpoint as @huangcq18 . This makes me think that this is not a bug in the code but something else that will be hard to find. Could be faulty hardware (supported by the observation that CPU inference works)

**Comment by ykirchhoff** [comment]
I could build a docker image and send it to you @RoloSmajda. That way we could rule out any software related issues. Does that make sense @FabianIsensee?

**Comment by RoloSmajda** [comment]
@FabianIsensee so i tried the `torch.any(torch.isnan(x))`  as you corrected me. the result is `tensor(False, device='cuda:0')` so no NaNs there. What I'm thinking right now is to try to run the prediction in Google Colab. I'll let you know if i was able to make it run in colab, and what was the result.

**Comment by RoloSmajda** [comment]
Okay so I managed to make it run in Colab. However, when the prediction is done, the cell keeps running, and in the output folder there is only json files, but not the predicted segmentation file. Maybe you guys have more insight on how to make it work in colab?

**Comment by FabianIsensee** [comment]
colab is a good idea!

**Comment by FabianIsensee** [comment]
Hey, I have no experience with colab. @ykirchhoff do you?

**Comment by ykirchhoff** [comment]
Not really much experience here either. But I think the hardware you get with colab is usually quite shitty regarding CPU and RAM, so maybe the segmentation export just takes forever. There is no way to check top/htop while a cell is running in the free version, right?

**Comment by FabianIsensee** [comment]
I'd be more optimistic in my formulation - after all this is a free service people can use to get started :-)
But yes it is unclear to me how CPU allocation works and to what extend we can actually rely on background workers. It could even be that you are awarded just one CPU core
